cranelift/wasm/src/environ/spec.rs

```diff
@@ -546,6 +546,53 @@ pub trait FuncEnvironment: TargetEnvironment {
         val: ir::Value,
     ) -> WasmResult<()>;
 
+    /// Translate an `i32.atomic.wait` WebAssembly instruction.
+    /// The `index` provided identifies the linear memory containing the value
+    /// to wait on, and `heap` is the heap reference returned by `make_heap`
+    /// for the same index.
+    ///
+    /// Returns an i32, which is negative if the helper call failed.
+    fn translate_atomic_wait_i32(
+        &mut self,
+        pos: FuncCursor,
+        index: MemoryIndex,
+        heap: ir::Heap,
+        addr: ir::Value,
+        expected: ir::Value,
+        timeout: ir::Value,
+    ) -> WasmResult<ir::Value>;
+
+    /// Translate an `i64.atomic.wait` WebAssembly instruction.
+    /// The `index` provided identifies the linear memory containing the value
+    /// to wait on, and `heap` is the heap reference returned by `make_heap`
+    /// for the same index.
+    ///
+    /// Returns an i64, which is negative if the helper call failed.
+    fn translate_atomic_wait_i64(
```

Comment:
In terms of API cleanliness, would it make sense to have a single function `translate_atomic_wait` that takes an extra Type parameter, that can be either I32 or I64? Going one step further, would it make sense to have a single `translate_atomic` method that would be passed an enum with variants `WaitI32/I64/Notify`?

Reply:
Merging the `wait` and `notify` cases is messy because they have different params.  However, the two `wait` cases can be merged without even needing an extra `Type` parameter, since the 32-vs-64 bit difference can be determined by inspecting the type of the `expected` parameter.  So that's what I did.